dra-evolution/pkg/api/quota_types.go

```diff
@@ -0,0 +1,57 @@
+package api
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Quota controls whether a ResourceClaim may get allocated.
+// Quota is namespaced and applies to claims within the same namespace.
+type Quota struct {
+	metav1.TypeMeta
+	// Standard object metadata.
+	metav1.ObjectMeta
+
+	Spec QuotaSpec
+}
+
+type QuotaSpec struct {
+	// AllowManagementAccess controls whether claims with ManagementAccess
+	// may be allocated. If multiple quota objects exist and at least one
+	// has a true value, access will be allowed. The default to deny such access.
+	//
+	// +optional
+	AllowManagementAccess bool `json:"allowManagementAccess,omitempty"`
+
+	// Stretch goals for 1.31:
+	//
+	// - maximum number of devices matching a selector
+	// - maximum sum of a certain quantity attribute
+	//
+	// These are additional requirements when checking whether a device
+	// instance can satisfy a request. Creating a claim is always allowed,
```

Comment:
I don't think it works well to have quota be asynchronous; it would work a lot better at admission time.

The problem is, if there are multiple ways to satisfy a claim, and quota is granular on a CEL-expression level, I don't think we can determine at admission time if the claim will violate quota. For example, if quota is `count(device.model == "H100") < 10`, and we have already consumed 9 H100s in this namespace, and we have claim that can be satisfied by A100 OR H100, then in theory the claim is satisfiable via A100s.

But this implies then that all schedulers and autoscalers become quota enforcement points, which I think is an expensive and risky proposition.

Either we need to simplify the claims or the quota system, I think.

Reply:
To make it work at admission time, we need to limit the expressiveness of quota to things that are semantically understood by the platform, AND that can be determined up front from the claim at admission time. Right now that just means:
* management access
* total count of devices

If we make driver required in the claim, rather than a CEL condition, then we can limit usage by driver. But we still can't by model.

Once allocatable device resources are part of the model, they can be used for quota.

We can't limit by model because it is part of the CEL expression, and I don't think we want to put something like that directly into the claim.

If we make class a required part of claim, then we could limit per class as well. I think that may be our best bet.  to make truly expressive quota that can be enforced at admission time. We should allow the admin to define the allow list for classes, plus numerical limits, and require class in the claim. That would give us the most flexibility here.

Reply:
> I don't think it works well to have quota be asynchronous; it would work a lot better at admission time.

Why? Better in which scenario?

It's also worse in some scenarios. For example, suppose someone uses Kubernetes as a batch system. It's normal to create more jobs than currently are allowed to run, so their claims will be "over quota". That's fine, jobs will run as soon as previous jobs are completed.

There may be some cases where a single claim is defined such that it can never be satisfied (asks for 100 devices, only 50 allowed in the namespace). While it would be nice to give users more immediate feedback when they do this, I don't think that it justifies replicating the allocation checks in the admission phase.

Management access is a bit different. Assuming that the namespace settings don't change, asking for it in a namespace where it's not allowed is a permanent failure. We could add an admission check for this.

I think this is again a "UX vs flexibility" question. You want a simple UX, but that can only work by not supporting more complex cases (over quota at admission time, under quota at allocation time). I prefer to start with a flexible approach and then add admission checks where they make sense later (management access).

Reply:
> It's also worse in some scenarios. For example, suppose someone uses Kubernetes as a batch system. It's normal to create more jobs than currently are allowed to run, so their claims will be "over quota". That's fine, jobs will run as soon as previous jobs are completed.

I don't think we should conflate "quota" - what is administratively allowed in this namespace - with "obtainability" - what devices are available for use. Quota for devices should work like other quota, and be applied at admission time. Whether it's memory or devices, as a workload author, I should be notified at admission time that I am asking for more than allowed.